libcloud/compute/drivers/cloudscale.py

```diff
@@ -0,0 +1,246 @@
+# limitations under the License.
+"""
+A driver for cloudscale.ch.
+"""
+
+import json
+
+from libcloud.utils.py3 import httplib
+
+from libcloud.common.base import ConnectionKey, JsonResponse
+from libcloud.compute.types import Provider, NodeState
+from libcloud.common.types import InvalidCredsError
+from libcloud.compute.base import NodeDriver
+from libcloud.compute.base import Node, NodeImage, NodeSize
+
+
+class CloudscaleResponse(JsonResponse):
+    valid_response_codes = [httplib.OK, httplib.ACCEPTED, httplib.CREATED,
+                            httplib.NO_CONTENT]
+
+    def parse_error(self):
+        body = self.parse_body()
+        if self.status == httplib.UNAUTHORIZED:
+            raise InvalidCredsError(body['detail'])
+        else:
+            # We are taking the first issue here. There might be multiple ones,
+            # but that doesn't really matter. It's nicer if the error is just
+            # one error (because it's a Python API and there's only one
+            # exception.
+            return next(iter(body.values()))
+
+    def success(self):
+        return self.status in self.valid_response_codes
+
+
+class CloudscaleConnection(ConnectionKey):
+    """
+    Connection class for the Vultr driver.
+    """
+    host = 'api.cloudscale.ch'
+    responseCls = CloudscaleResponse
+
+    def add_default_headers(self, headers):
+        """
+        Add headers that are necessary for every request
+
+        This method adds ``token`` to the request.
+        """
+        headers['Authorization'] = 'Bearer %s' % (self.key)
+        headers['Content-Type'] = 'application/json'
+        return headers
+
+
+class CloudscaleNodeDriver(NodeDriver):
+    """
+    Cloudscale's node driver.
+    """
+
+    connectionCls = CloudscaleConnection
+
+    type = Provider.CLOUDSCALE
+    name = 'Cloudscale'
+    website = 'https://www.cloudscale.ch'
+
+    NODE_STATE_MAP = dict(
+        changing=NodeState.PENDING,
+        running=NodeState.RUNNING,
+        stopped=NodeState.STOPPED,
+        paused=NodeState.PAUSED,
+    )
+
+    def __init__(self, key, **kwargs):
+        super(CloudscaleNodeDriver, self).__init__(key, **kwargs)
+
+    def list_nodes(self):
+        '''
+        List all your existing compute nodes.
+        '''
+        return self._list_resources('/v1/servers', self._to_node)
+
+    def list_sizes(self):
+        '''
+        Lists all available sizes. On cloudscale these are known as flavors.
+        '''
+        return self._list_resources('/v1/flavors', self._to_size)
+
+    def list_images(self):
+        '''
+        List all images.
+
+        Images are identified by slugs on cloudscale.ch. This means that minor
+        version upgrades (e.g. Ubuntu 16.04.1 to Ubuntu 16.04.2) will be
+        possible within the same id ``ubuntu-16.04``.
+        '''
+        return self._list_resources('/v1/images', self._to_image)
+
+    def create_node(self, name, size, image, location=None, ex_create_attr={}):
```

Comment:
I would argue that some other drivers (e.g. digitalocean) are doing it the same way. Prefixing everything with `ex_` feels clunky and having a dictionary makes it easy to be "feature-compatible".